File: sudoku2rm/renderer.py
import io

from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import mm
from reportlab.pdfgen import canvas

# reMarkable 2 display is ~157mm x 210mm — A4 fits perfectly with minimal scrolling.
# Grid: 180mm × 180mm → cells of 20mm each, comfortable for handwriting.
_GRID_MM = 180
_FONT_GIVEN = "Helvetica-Bold"
_FONT_SIZE = 22
_LINE_THIN = 0.6
_LINE_THICK = 2.5


def render_sudoku_to_pdf(grid: list[list[int]], title: str = "Sudoku") -> bytes:
    """Render a parsed sudoku grid to a reMarkable-optimised A4 PDF.

    Pre-filled (given) numbers are printed in bold; empty cells are blank for
    handwriting. Returns the PDF as raw bytes.
    """
    buf = io.BytesIO()
    page_w, page_h = A4

    grid_size = _GRID_MM * mm
    cell = grid_size / 9

    # Centre grid horizontally; leave 30 mm at top for title
    gx = (page_w - grid_size) / 2          # grid left edge
    gy = page_h - 30 * mm - grid_size      # grid bottom edge (PDF y is from bottom)

    c = canvas.Canvas(buf, pagesize=A4)

    # ── Title ──────────────────────────────────────────────────────────────
    c.setFont(_FONT_GIVEN, 26)
    tw = c.stringWidth(title, _FONT_GIVEN, 26)
    c.drawString((page_w - tw) / 2, page_h - 22 * mm, title)

    # ── Given numbers ──────────────────────────────────────────────────────
    c.setFont(_FONT_GIVEN, _FONT_SIZE)
    c.setFillColor(colors.black)
    for row in range(9):
        for col in range(9):
            val = grid[row][col]
            if val == 0:
                continue
            # cell bottom-left in PDF coords (row 0 = top of grid)
            cx = gx + col * cell
            cy = gy + (8 - row) * cell
            text = str(val)
            tw = c.stringWidth(text, _FONT_GIVEN, _FONT_SIZE)
            # Horizontally centred; baseline ~33 % up the cell
            c.drawString(cx + (cell - tw) / 2, cy + cell * 0.33, text)

    # ── Thin cell lines (skip box boundaries — drawn thick below) ──────────
    c.setStrokeColor(colors.black)
    c.setLineWidth(_LINE_THIN)
    for i in range(1, 9):
        if i % 3 == 0:
            continue
        x = gx + i * cell
        c.line(x, gy, x, gy + grid_size)           # vertical
        y = gy + i * cell
        c.line(gx, y, gx + grid_size, y)           # horizontal

    # ── Thick box lines (at 0, 3, 6, 9 cell boundaries) ───────────────────
    c.setLineWidth(_LINE_THICK)
    for i in range(4):
        x = gx + i * 3 * cell
        c.line(x, gy, x, gy + grid_size)
        y = gy + i * 3 * cell
        c.line(gx, y, gx + grid_size, y)

    c.save()
    buf.seek(0)
    return buf.read()

File: sudoku-web/lib/renderer.ts
import { PDFDocument, StandardFonts, rgb } from "pdf-lib";

// A4 in points (1pt = 1/72 inch)
const PAGE_W = 595.28;
const PAGE_H = 841.89;
// 180mm grid → 510 pts; each cell = 56.7 pts (~20mm, comfortable for handwriting)
const GRID_SIZE = 510;
const CELL = GRID_SIZE / 9;

export async function renderSudokuToPdf(
  grid: number[][],
  title: string
): Promise<Uint8Array> {
  const doc = await PDFDocument.create();
  const page = doc.addPage([PAGE_W, PAGE_H]);
  const font = await doc.embedFont(StandardFonts.HelveticaBold);

  // grid bottom-left: centred horizontally, 30mm (85 pt) from top
  const gx = (PAGE_W - GRID_SIZE) / 2;
  const gy = PAGE_H - 85 - GRID_SIZE;

  // Title
  const titleSize = 26;
  const titleW = font.widthOfTextAtSize(title, titleSize);
  page.drawText(title, {
    x: (PAGE_W - titleW) / 2,
    y: PAGE_H - 62,
    font,
    size: titleSize,
    color: rgb(0, 0, 0),
  });

  // Given numbers
  const numSize = 22;
  for (let row = 0; row < 9; row++) {
    for (let col = 0; col < 9; col++) {
      const val = grid[row][col];
      if (!val) continue;
      const text = String(val);
      const cx = gx + col * CELL;
      const cy = gy + (8 - row) * CELL;
      const tw = font.widthOfTextAtSize(text, numSize);
      page.drawText(text, {
        x: cx + (CELL - tw) / 2,
        y: cy + CELL * 0.33,
        font,
        size: numSize,
        color: rgb(0, 0, 0),
      });
    }
  }

  // Thin cell lines
  for (let i = 1; i < 9; i++) {
    if (i % 3 === 0) continue;
    const x = gx + i * CELL;
    page.drawLine({ start: { x, y: gy }, end: { x, y: gy + GRID_SIZE }, thickness: 0.6, color: rgb(0, 0, 0) });
    const y = gy + i * CELL;
    page.drawLine({ start: { x: gx, y }, end: { x: gx + GRID_SIZE, y }, thickness: 0.6, color: rgb(0, 0, 0) });
  }

  // Thick box lines (every 3 cells)
  for (let i = 0; i <= 3; i++) {
    const x = gx + i * 3 * CELL;
    page.drawLine({ start: { x, y: gy }, end: { x, y: gy + GRID_SIZE }, thickness: 2.5, color: rgb(0, 0, 0) });
    const y = gy + i * 3 * CELL;
    page.drawLine({ start: { x: gx, y }, end: { x: gx + GRID_SIZE, y }, thickness: 2.5, color: rgb(0, 0, 0) });
  }

  return doc.save();
}
